### DISCUSSION TITLE
Defining Desk-lang

### DISCUSSION THREAD

**Discussion by hsk** [Hihaheho/Desk#76] [Ideas]
日本語で大丈夫ですか？

**Comment by hsk** [comment]
反応がないのですけどどうしたら良いでしょうか？？？

**Reply by ryo33** [reply]
日本語で大丈夫です！エフェクトシステムとユニオン型の文法と型推論規則、評価規則を準備中です。型推論規則についてはHindley-Milnerではないのでちょっと反則的な書き方になる可能性がありますが、おそらく杞憂です。いつになるかは約束できないです。

**Comment by ryo33** [comment]
I've written the syntax of `term`, but I don't know a nice OCR software for it, so I post it here as is.![image](https://user-images.githubusercontent.com/8780513/227508229-3c038adb-94b9-470d-b278-a1efc9d48a92.jpeg)

**Reply by ryo33** [reply]
実用的にすることを考えると繰り返しの中ではカンマをオプショナルで使えるようにしたいです。

**Comment by hsk** [comment]
```ebnf
{T}  : empty or more repetitions separated by optional commas (trailing comma is valid)

term = NAT                                               // natural number
        | "<" { term } ">"                                   // prod
        | "[" { term } "]"                                    // vec
        | "{" { term "=>" term } "}"                          // map
        | "λ" type "->" term                                // function
        | "@" IDENT term                                    // label
        | "@@" IDENT term                                   // brand
        | "branch" term "{" { type "=>" term } "}"          // branch
        | "let" term "in" term                              // let
        | "&" type { term }                                   // ref
        | "!" term "~>" type                                // perform
        | "handle" term "{" {type "~>" type "=>" term} "}"  // handle
        | term ":" type                                     // type annotation
        | "Π" term                                          // dependent product
        | "Σ" term                                          // dependent sum
        | "#" DSON term                                     // attr
```

**Reply by ryo33** [reply]
ありがとうございます！！！
次回はもっときれいに書くか、自分で書き起こすかするつもりです。
勝手に編集して、手書きで不満に思っていた部分などのtrivialな変更を行いました。文法は変わっていません。

**Comment by hsk** [comment]
これはDesk言語の構文ですか？

**Reply by ryo33** [reply]
はい現在の実装とは乖離がありますが定義にあわせるかたちで将来的に統一します

**Comment by hsk** [comment]
小さい言語を作るというような話もありましたので確認させてもらいました。
ありがとうございます。

**Comment by hsk** [comment]
主にTAPLやCoPLなどOCaml勢の慣習にしたがって構文を書き直すと以下のように書けます:

```
Syntax
τ ::=                                                   type
t ::=                                                   term
    | n                                                 natural number
    | [ t1, t2, t3, ..., tn ]                           vec
    | { t11 => t12, t21 => t22, ..., tn1 => tn2 }       map
    | λ τ -> t                                          function
    | @IDENT t                                          label
    | branch t { τ11 => t12,τ21 => t22,...,τn1 => tn2 } branch
    | let t in t                                        let
    | &τ t1,t2,t3,...,tn                                ref
    | !t ~> τ                                           perform
    | handle t { τ11~>τ12=>t13, τ21~>τ22=>t23, ..., τn1~>τn2=>tn3 }
                                                        handle
    | t : τ                                             type annotation
    | @@IDENT t                                         brand
    | #DSON t                                           attr
    | <t1, t2, t3, ..., tn>                             prod
    | Π t                                               dependent product
    | Σ t                                               dependent sum
```

非終端記号は１文字で表し、繰り返しはt1,t2,t3,...,tnのように書いて0個からか1個からかまで厳密には表スことよりわかりやすさを優先します。term は t、type も t のように頭文字が重なってしまう場合はtype を τ のようにギリシャ文字を使うこともよくあります。

**Reply by ryo33** [reply]
確かにそちらのほうがわかりやすいです！今後はそちらでいきます。文法はEBNFなのかABNFなのか何を使うのがいいか見当がついていません。

**Reply by hsk** [reply]
構文解析をすることを考えるのであればEBNFなどを使うと良いと思いますが、論文で用いられるような形式手法の形で書く場合は Rust の enum を書くようなイメージで構文を定義するのでこのような形で書くことになっていると私は思います。
構文解析も作れますけどパーサのことは考えずにASTを考えるようなつもりで書いています。

**Reply by hsk** [reply]
自分の場合はtとτがかぶるのが嫌なのでtermと考えずにexpressionと考えてtをe,τをtと書く事が多いです。
研究者によってはtermのmをとってterm を m、type を t と書くこともありけっこう自由に書いて良いものだと思うようになると思います。
Rustのenumの名前は好きにつけるのと同じです。数学的に手書きすることを考えると一文字で書くのが普通のプログラムと違うところでしょう。
名前は型も同時に表すので慣れると考えるのが楽になります。

**Reply by ryo33** [reply]
ありがとうございます！自分で使っていた形式は個人的に書きにくいのと読みづらいので慣習からあまり離れない範囲でDesk言語にあった書き方を考えてみようと思います。

**Reply by ryo33** [reply]
```
項::=
| λ τ → 項 (function)
| Nat
| @ Ident 項 (label)
| @@ Ident 項 (brand)
```
概ねこれで行こうかと

**Reply by hsk** [reply]
普通のプログラマとしては１文字で項を表すことに抵抗があるので１文字で書くなら漢字にさせてくれ。
コメントを突然書くと気持ち悪いからカッコ内にコメントを書くことにさせてくれ。みたいな気持ちなのはわかりました。

**Comment by ryo33** [comment]
実は私が書いたProductの文法には強い不満があります。ほかに良さげなものを思いつかなかったので、仮のものを書いた感じです。
現状の実装では、項と型ともに`"*" "<" { term/type } ">"`となっていますが、それも微妙だと思っています。
要件は
- 0-or-more な要素を含むことができる
- product/sumで相似である
- 数学的な率直であるがゆえの美しさをもつ

あたりになります。
`A x B x C`のようにして別に`0`や`⊤`を作ることが一番ということは分かっているのですが、構文を過剰に減らしたいのと、ユーザーに「空のProduct」という存在を実体験してほしいため上記のような思想になっています。

**Comment by hsk** [comment]
Desk言語の実装に置いては細かい構文は重要になると思いますが、論文では主に意味論などに興味があるためproduct の機能を導入するならproduct < t1/τ1,t2/τ2,...,tn/τn > とそのまま書くほうがわかりやすいと思います。

**Reply by ryo33** [reply]
了解です。論文と実装で文法をできるだけ対応させたい思いがあったのでそのような方針になっていました。できるだけそのまま理解できるもので納得できる構文を考えてみます。

**Comment by hsk** [comment]
productの意味がよくわからないのですけどtupleな感じなのか、レコードな感じなのか、effect systemの話なのかがよくわからないので例があるとよいのではないかと思います。

**Reply by ryo33** [reply]
型を順序のない集合とした時のプロダクトになります。どう例を出していいか見当がついていません

**Reply by hsk** [reply]
なるほど、JavaやScalaでいうとSetのことですね。

**Reply by hsk** [reply]
<> + 1 = <1>
<> + 1 + 2 = <1,2>
<> + 1 + 2 + 1 = <1,2>
<> + 3 + 2 + 1 = <1,2,3>
<2,3> + <3,4> = <2,3,4>
<1,2,3> - <2,3,5> = <1>
みたいな計算ができるのではないでしょうか？

**Reply by ryo33** [reply]
はいかなり近いです。Productについては実際はkeyが型のレコードになっていて、要素が値になります。同じキーがある場合は後者の値が使われます。キーを減らす時は型強制をおこないます。
Sumについても型がタグになったタグ付きユニオンです。sakuraiさんの書かれたように同じ型があれば一つに潰れ、型強制により型レベルで要素を増やせます。項レベルではリテラルがなく、branch構文とhandle構文の評価でしか構築できません。また場合分けで要素を取り出すのはbranch構文によってのみ行えます。

**Reply by hsk** [reply]
じゃあレコードじゃないかと思いますけど

<> + <x:1> = <x:1>
<> + <x:1> + <y:2> = <x:1,y:2>
<> + <x:1> + <y:2> + <x:3> = <x:3,y:2>
<> + <z:1> + <y:2> + <x:3> = <x:3,y:2,z:1>
<x:1,y:2> + <y:3,z:4> = <x:1,y:3,z:4>
<x:1,y:2,z:3> - <y:2,z:3,a:5> = <x:1>
みたいにつかえるmapなのかな。

**Comment by ryo33** [comment]
決めました。termという概念を消します。そうすれば私がDesk言語の文法や型システムに対して抱いていた色々な不満を一挙に解決できそうです。
明日文法定義からやり直します。

**Reply by ryo33** [reply]
消すというかterm = typeにします

**Comment by hsk** [comment]
いくつか ICFP などの論文の構文定義の例を貼りますので参考にしてください。

<img width="584" alt="スクリーンショット 2023-03-25 7 56 10" src="https://user-images.githubusercontent.com/129693/227658187-b7bbffb2-5303-4887-9861-8b959dd154c3.png">

Scala3のDOT計算の論文 System D<: の構文

https://namin.seas.harvard.edu/files/dot_wadlerfest.pdf

**Comment by hsk** [comment]
<img width="559" alt="スクリーンショット 2023-03-25 8 02 33" src="https://user-images.githubusercontent.com/129693/227658864-582b3564-4f33-4b5e-b702-cea26b5ab734.png">

effect system の単純な実装 λeff の構文

https://github.com/Nymphium/lambdaeff/blob/master/text/main.pdf

**Comment by hsk** [comment]
Rust関連の論文の構文
<img width="466" alt="スクリーンショット 2023-03-25 8 04 21" src="https://user-images.githubusercontent.com/129693/227659060-78828973-ef3f-48c5-a9ae-6445363e78b4.png">


https://arxiv.org/abs/2206.07185

**Comment by hsk** [comment]
ICFP21の代数的エフェクトの論文の構文

<img width="677" alt="スクリーンショット 2023-03-25 8 08 13" src="https://user-images.githubusercontent.com/129693/227659403-5443880a-59e5-4a9d-bd40-a493f3a88f0c.png">

https://dl.acm.org/doi/10.1145/3473580

**Comment by hsk** [comment]
Automatically deriving control-flow graph generators from operational semantics

<img width="303" alt="スクリーンショット 2023-03-25 8 27 31" src="https://user-images.githubusercontent.com/129693/227661056-4ebee6b1-c56d-4aa7-b783-ff0f3b569ec4.png">

https://dl.acm.org/doi/10.1145/3547648

こういうパターンかｗ

**Comment by hsk** [comment]
多相レコード型に基づくRubyプログラムの型推論

<img width="696" alt="スクリーンショット 2023-03-25 8 38 48" src="https://user-images.githubusercontent.com/129693/227662008-dc6f408a-f846-4697-b230-0dd467a9a081.png">
<img width="299" alt="スクリーンショット 2023-03-25 8 39 06" src="https://user-images.githubusercontent.com/129693/227662034-400955d5-f85c-49a9-bab0-d763cbe2d141.png">

https://cir.nii.ac.jp/crid/1050282812867199616

**Comment by hsk** [comment]
色々な言語の論文の構文定義を拾ってきて見ましたが、よくあるパターンだと構文のコメントはカッコで括らなくて、１文字で非終端記号を表してます。ただカッコでコメントを書く人もいる。コメントは書かずに説明するパターンも多々ある感じです。

**Comment by ryo33** [comment]
文法を更新しました。些細なミスとかはあるかもしれませんが内容としては満足いくものになりました。
https://github.com/Hihaheho/Desk/blob/main/docs/language/formal-definition.md

**Reply by hsk** [reply]
いい感じですね。その調子で慣れていけば良いと思います。

いくつか気になった点があったので修正してみました：

```
tn::=                                          (number type)
    | Nat | Int | Rat | Real
    | uN | iN                                  (n-bit integer)
tt::=
    | U                                        (universe)
    | { t1, ... }                              (unordered set of types)
f ::=
    `tt` or super set of t with any t in it is replaced to `t | tn` (type family)
t ::=
    | tn
    | tn < t
    | tn < t
    | ' Char '
    | Π f
    | Σ f
    | λ t → t
    | Ident                                     (variable)
    | [ t1, ... ]                               (fixed-length vector)
    | [ t; t_len ]                              (free-length vector)
    | < t_key1 => t_value1, ... >               (map)
    | let t_def in t
    | & t t1, ...                               (application)
    | branch t { t_case1 => t1, ... }
    | ! t ~> t_out                              (perform)
    | handle t { t_in1 ~> t_out1 => t1, ... }
    | @ Ident t                                 (label)
    | @@ Ident t                                (brand)
    | t_from is t_to                            (exploit type coercion)
    | ?                                         (hole)
    | # Dson t                                  (attribute)
```

:== ではなく ::= にしましょう。
コメントは左揃え過右揃えにすると良いです。面倒でしょうがコメントを付けると決めたら全部つけましょう。
特に非終端記号というか型名にはコメントを付けたほうが良いと思います。

**Reply by ryo33** [reply]
:== は恥ずかしいミスでした。

コメントなのですが、self-descriptiveなものに説明がついていると自分の感覚では逆に分かりづらくなるのでこのようになっています。コメントがあったりなかったりするのが美しくないのはわかっています。self-descriptiveな構文で英語の代わりに記号を使うようにするのがいいのかもしれません。
インデントについてはあった方がいいと思っているのですが、コメントなどを揃えるのは個人的には好みではないです。vimであってもプラグインを使わないと不便ですし、モバイル端末から書いたりすることもあるので難しいと感じています。ただ、確かに読みづらいということは分かっているので何か解決方法を探してみます。mathjaxを使えば一発だと思いますが、手書きするのは苦手なのでchatgptとかにやってもらうのがいいのかもしれません。

**Reply by hsk** [reply]
コメントには説明というか名前をつけるのが普通で名前を全くつけないこともあります。
インデントを付けないで書くようなパターンを見たことはないです。
プログラミング言語でコメントを付ける場合の間隔でコメントを書きたいということでカッコをコメントにするというルールを決めたのでしょうけど、慣習的にそういった流儀で書く人は見たことがないので汚く感じます。
個人の好みでどうのという話ではないかな。
コメントを書くなら書く、書かないなら一切書かないで、説明するのが普通だと思います。

**Reply by ryo33** [reply]
確かに後から評価規則などを書くわけですしコメントなくてもいいですよね。

**Reply by hsk** [reply]
最初は練習に書いたほうがいいんじゃないかと思いますけど、量が多いから書くのがめんどくさいのでしょうｗ

**Comment by ryo33** [comment]
コンテキストも書きました。あとで追記します。
```
Γ :== 
| • (empty)
| Γ, a (beginning of a scope of a type variable)
| Γ, x: t (type conclusion of variable x)
| Γ, Σ f (beginning of a scope of an existential type)
| Γ, Σ f: t (type conclusion of an existential type)
| Γ, ^a (introducing of a marker)
```

**Comment by ryo33** [comment]
型システムの方も書き進めています。
![image](https://user-images.githubusercontent.com/8780513/227667720-41d22043-d1a5-4617-9150-5fd1bdbfb363.jpeg)

**Reply by ryo33** [reply]
mathjaxとかで書き起こすの面倒くさくてモチベーションがないので、prolog的な感じで書きおこすかもです。

**Reply by ryo33** [reply]
ProductとSumのサブタイピング規則です。

![image](https://user-images.githubusercontent.com/8780513/227675737-620af2ef-df2a-4574-aa33-a5e191b77b8a.jpeg)

**Reply by ryo33** [reply]
@hsk 上記についてどう思われますか？型付け規則と評価規則がないので包括的ではないですが、ここの規則がコアなので

**Reply by hsk** [reply]
構文の意味がわからない状態で部分型付けの図が突然でてきているのでパット見はよくわかりません。
{}で２つの式が下に書いてあるのがまた意味がよくわかりません。
この図に自然言語で説明があればわかるかもしれませんし、わからないかも知れません。
４つか６つの規則で計算できるなら楽で良いなと思います。

```
t = tx ∧ t != tn != x
-------------------
t <:Ω tx

@x t ∧ @x tx ∧ t <:Ω tx
-------------------------
t <:Ω tx
```

この２つの規則の tx が何なのかよくわからないです。t と tx は t1 と t2 と書き換えられるものなのか、書き換えられないものなのかわかりません。
まず構文の説明がないので現段階で理解するのが難しいのは仕方ないかと思います。

**Reply by hsk** [reply]
```
Γi |- ta <:Ω tb -| Γo
-----------------------------------
Γi |- ta <: Σ{tb1,...,tbn} -| Γo

Γi |- ta <:Ω tb -| Γo
-----------------------------------
Γi |- Π{ta, ...} <: tb -| Γo

Γi |- ta1 <:Ω tb1 -| Γ,... Γ |- tan <:Ω tbn -| Γ
----------------------------------------------------
Γi |- Σ{ta1,...,tan} <: Σ{tb1,...,tbn,tc1,...} -| Γ

Γi |- ta1 <:Ω tb1 -| Γ,... Γ |- tan <:Ω tbn -| Γ
----------------------------------------------------
Γi |- Π{ta1,...,tan,tc1,...} <: Π{tb1,...,tbn} -| Γ
```

ってことですかね？

**Reply by hsk** [reply]
```
Γ ::= 
| •         (空)
| Γ, a      (型変数のスコープの先頭。)
| Γ, x: t   (変数 xの型)
| Γ, Σ f    (存在型スコープの先頭。)
| Γ, Σ f: t (存在型 f の型)
| Γ, ^a     (マーカー導入)
```

というような話なのかな？

```

Γinput |- Predicate -| Γoutput

τ1 <: τ2 is reflxsive subtype
Σf <:= τ  is instantiation

τ1 should τi   <-<-  <-: これ何？
τ2 is     τ2   is<-  :->: is->
```


```
---------------------------- (Sub-Id)
Γ1,a,Γ2 |- a <: a -| Γ1,a,Γ2
```

これは a は a の型ということですよね。
恐らく a が変数であり、aが環境Γ内にあるならば、a は a のサブタイプだということでしょう。

こう書いたほうがわかり良い気がします:

```
a ∈ Γ
---------------------------- (Sub-Id)
Γ |- a <: a -| Γ
```

**Reply by hsk** [reply]
```
-------------------------------- (Sub-ExistId)
Γ1,Σf,Γ2 |- Σf <: Σf -| Γ1,Σf,Γ2
```

これも存在型の同じやつという話だろうな。

```
Σf ∈ Γ
---------------------------- (Sub-ExistId)
Γ |- Σf <: Σf -| Γ
```

と書いたほうがわかり良い気がする。


```
-------------------------------- (Sub-Number)
Γ |- nat <: int <: rat <: real -| Γ
```

これもいいたいことはわかるけど、

```
-------------------------------- (Sub-NatInt)
Γ |- nat <: int -| Γ

-------------------------------- (Sub-IntRat)
Γ |- int <: rat -| Γ

-------------------------------- (Sub-RatReal)
Γ |- rat <: real -| Γ
```

と分けたほうがわかりやすいかな。

**Reply by hsk** [reply]
```
Γ1 |- a <: b -| Γ2
Γ2 |- b <: c -| Γ3
-------------------------------- (Sub-Trans)
Γ1 |- a <: c -| Γ3
```

推移即があると、宣言的になってしまうのでアルゴリズミックではなくなるけど、とりあえず数のサブタイプ関係は説明できますね。

**Reply by hsk** [reply]
```
Γ1 |- A1 <: A2 -| Γ2
Γ2 |- B1 <: B2 -| Γ3
---------------------------------- (Sub-Fun)
Γ1 |- A1 -> B1 <: A2 ->B2 -| Γ3
```

関数のサブタイプ関係でしょう。


```
Γ1 |- B <: A -| Γ2
---------------------------------- (Sub-???)
Γ1 |- A[] <: B[] -| Γ2
```

よくわかりませんけど、ひっくり返るらしい。こんな定義あったっけ？


```
Γ1 |- B1 <: A1 -| Γ2
Γ2 |- A2 <: B2 -| Γ3
--------------------------------------
Γ1 |- A1 => B1 <: A2 <: B2 -| Γ3
```

=> ってなんだっけ？突然A2がでてきたのだけど、なぜA2とB2とA1とB1が関係するようになったのか謎。


```
Γ1 |- τ1 is τ2 -| Γ2
----------------------------------- (E-Let)
Γ1 |- let τ3 in τ2 is τ2 -| Γ2
```

Let 式の評価規則か何かだろうか？

**Reply by ryo33** [reply]
ありがとうございます。概ね書いていただいた内容で意図通りです。

https://github.com/Hihaheho/Desk/discussions/76#discussioncomment-5423900
Γの文法については先にこちらで書いていました。
isとshouldについては説明し忘れていました。
`t is τ`は慣例に習って`t: τ`と書くべきだったかもしれません。`t should τ`は規則の上側にある`t: τ`に近いニュアンスですが、`should`についての規則群が別に定義されるので概念としては別のものです。引用し忘れたのですがこちらの論文[Complete and Easy Bidirectional Typechecking for Higher-Rank Polymorphism](https://arxiv.org/abs/1306.6032)を参照していてそのなかではそれぞれ`=>`と`<=`になっています。
`Sub-Id`や`Sub-Exist`についてはこの定義でのΓが集合ではなく順序のある列であるため、私が書いたもののように書く必要があります。

**Reply by ryo33** [reply]
> 構文の意味がわからない状態で部分型付けの図が突然でてきているのでパット見はよくわかりません。 {}で２つの式が下に書いてあるのがまた意味がよくわかりません。 この図に自然言語で説明があればわかるかもしれませんし、わからないかも知れません。 ４つか６つの規則で計算できるなら楽で良いなと思います。
> 
> ```
> t = tx ∧ t != tn != x
> -------------------
> t <:Ω tx
> 
> @x t ∧ @x tx ∧ t <:Ω tx
> -------------------------
> t <:Ω tx
> ```
> 
> この２つの規則の tx が何なのかよくわからないです。t と tx は t1 と t2 と書き換えられるものなのか、書き換えられないものなのかわかりません。 まず構文の説明がないので現段階で理解するのが難しいのは仕方ないかと思います。

これについてはまず関係のない２つのものに`x`を使ってしまっていたのが大誤算でした。正しくは以下のように書くべきでした。
```
t = ta ∧ ∀tn where n != a. t != tn
---------------------------------------------
t <:Ω ta
```

```
(`ta`以外に`@ x t ∧ @ x tn`がない) ∧ t <:Ω ta
------------------------------------------------------------
@ x t <:Ω @ x ta
```
記法に自身がないので正しい書き方が有ればおしえてください。

**Reply by hsk** [reply]
Γの要素を取り出すという話でリストを環境としてつかってスタックのように使う技法が関数型言語やPrologで一般的によく使われます。

prolog

```prolog
:- Γ=[x=int,y=int,x=bool],member(x=T,Γ),T=int
```
OCamlであれば、

```ocaml
List.assoc "x" [("x",1);("y",2);("x",3)];;
- : int = 1
```

のように書けて集合のように扱えるので、x ∈ Γ とかいても意味が通じます。

```
ーーーー
x ∈ x,Γ 

x ∈ Γ 
ーーーー
x ∈ _,Γ 
```

このように定義すればいいわけです。

prologのmember/2の定義はこちら:
```prolog
ーーーー
member(X,[X|_]) 

member(X,Γ) 
ーーーー
member(X,[_|Γ]) 
```

**Reply by ryo33** [reply]
@hsk 上記の回答はいかがでしょうか？
<:Ωは例えばΣ{A,B}をΣ{C,D}にサブタイピングを行う際に型同士の変換距離を使って一意に写像を確定させるための関係で、その対応が最も距離が近く、他に同じ距離の対応が存在しないことを表します。２つだけ規則を書きましたがもしかしたら今後の思いつきで数個増えるかもしれません。

**Reply by hsk** [reply]
> これについてはまず関係のない２つのものにxを使ってしまっていたのが大誤算でした。正しくは以下のように書くべきでした。
> ```
> t = ta ∧ ∀tn where n != a. t != tn
> ---------------------------------------------
> t <:Ω ta
> ```
>
> ```
> (`ta`以外に`@ x t ∧ @ x tn`がない) ∧ t <:Ω ta
> ------------------------------------------------------------
> @ x t <:Ω @ x ta
> ```
> 記法に自身がないので正しい書き方が有ればおしえてください。

```
B
------
A
```

はまず、BならばAとよみます。A where B ともかけます。条件が２つある場合は

```
B    C
------
A
```
とかいて、B かつ C ならば A とよみます。 A where B, C みたいに書くこともあります。
ですので、ーーーーを使って書くのであればwhereを使う必要はなくかけます。
またーーーーを使った記法では変数xは一階述語論理の全称変数∀x のことになるので必要がなければ∀の記号は使いません。

t が項であることはわかるのですが ta がなになのかわかりかねます。taのaが変数ということなのであれば、aをわざわざ書かずに

t1 と t2 についてかけばいいだけのように思えますが、ta の 型がどこからやってくるのかわかりません。
恐らく基底型のようなかたちでnatやintなどがあってそれをtaと書いているのかも知れませんがよくわからないです。

number ::= nat | int | rat | real
b ::= number | String
のように書いて参照するとわかりやすくなるかと思います。

**Reply by hsk** [reply]
まず、Σ{A,B}をΣ{C,D}のΣの意味がよくわからない状態で色々書かれても、、、。というところですのでまず構文の定義をしっかりしてから部分型付けの説明をしていただけると議論がしやすいのではないかと思います。

**Reply by hsk** [reply]
Ryoさんは https://zenn.dev/ryo33/articles/4b11a0dfede4c5 を読んで Bidirectional Type-checking を理解していることを前提で話しているようなのでその辺読んでみますね。

**Reply by ryo33** [reply]
ありがとうございます。かなり勉強になってます。taはt1~tnの一つの要素になります。
latex的にt_aと書くのがわかりやすかったかもしれません。
確かに∈を順序集合に使ってはいけないというようなルールはありませんし、Γの定義にもよるのですね。
そちらの論文だと`Γ1, x, Γ2`を`Γ[x]`と表現しているのですが意味が直感的でないと思ったのと、他の規則ではΓ1とΓ2の両方が規則中に使用されることが多いので記法に一貫性を持たせる意図がありました。

**Reply by ryo33** [reply]
ΣとΠは集合の総和と総積です。

**Reply by ryo33** [reply]
あと私が「どうですか？」的なふんわりとした質問をする時は「記法は変じゃないですか？」とか（　「なんとなく言いたいことは通じますか？」的なニュアンスです。定義が進んできて定義の詳細について議論をしたい場合は具体的な質問を行いたいと考えています。

**Comment by hsk** [comment]
型付け規則や評価規則、部分型付け規則も様々な流儀があります。

評価規則 |- e --> v

```
---------- (E-Int)
|- i --> i

|- e1 --> i1    |- e2 --> i2    i1 + i2 is i
-------------------------------------------- (E-Plus)
|- e1+e2 --> i
```

評価規則 e --> v

```
---------- (E-Int)
i --> i

e1 --> i1    e2 --> i2    i1 + i2 is i
-------------------------------------------- (E-Plus)
e1+e2 --> i
```

評価規則 {e} --> {v}    (前提をifで書く)

```
(E-Int)     {i}     --> {i} 
(E-Plus)    {e1+e2} --> {i}   if {e1} --> {i1}, {e2} --> {i2}, i1 + i2 is i
```

評価規則 {e} --> {v}     (前提をiffで書く)

```
(E-Int)     {i}     --> {i} 
(E-Plus)    {e1+e2} --> {i}   iff {e1} --> {i1}, {e2} --> {i2}, i1 + i2 is i
```

評価規則 e ⇓ v     (前提をwhereで書く)

```
E-Int     i     ⇓ i 
E-Plus    e1+e2 ⇓ i   where e1 ⇓ i1, e2 ⇓ i2, i1 + i2 is i
```

要はわかれば良いのですが、ネストを合わせるときれいに見えます。

**Reply by ryo33** [reply]
1番上が見慣れてるのでそれで進めるつもりです。

**Comment by hsk** [comment]
Desk言語は、普通の言語からかなりかけ離れた思想で作られていると思います。
代数的効果もかなり新しい考え方ですので、詳しい説明がないとわかりません。
let式やλ抽象にしても普通は let x = 1 in x のように書いたり (λx.x+1) 3 のように使うわけですが、Deskは変数名がないので、普通ではないですよね。
その辺の説明がない状態で型付け規則を書かれてもチョットマッテとなります。

まずは構文を書いて、各要素に名前をつけそれぞれの機能についての使い方の例をあげて分かったとなった後に型付け規則などで説明してもらえるとわかりやすくなると思います。
型付け規則や、評価規則はたくさん書くことになり、それぞれの意味を説明したり証明で参照するために名前をつけると良いようです。それぞれの意味を図の説明として書くと良いと思います。

**Reply by ryo33** [reply]
あれもしかして評価規則のほうを先に書いたほうがコラボレーションとしては進めやすいんですかね？
そうですね規則にはちゃんとなまえをつけるようにしようと思います。

**Reply by ryo33** [reply]
ゼミで指導を受けているような感覚で私としてはすごくやりやすいのですが、sakuraiさんにとってもどかしい状態になっていないか不安です。
すべてを書ききる前に細かくフィードバックをもらったほうがいいなと個人的に思っていたので、清書することなく書いた瞬間に投稿していましたが、もしサステナブルでないのであれば、もっと大きな粒度で書き起こししたものを投稿する方向にしたほうがいいのかもしれません。

**Reply by hsk** [reply]
俺は大丈夫だけど、Ryoさんが嫌になってやめないか不安ですｗ

**Reply by hsk** [reply]
とりあえずは、構文書いて使い方の例とかを書いてドキュメント整備してから、型付け関係などのはなしに入ると良いと思います。

**Reply by ryo33** [reply]
良かったです。私は、今後も続けると思いますが、やりたくないことはやらない方針で活動しているので、虫食い算的なものが出来上がらないのか不安ではあります。

**Reply by hsk** [reply]
ま、辛いこともあるかと思いますが、できるだけ楽しくやりましょう！

**Comment by hsk** [comment]
部分型付けの規則は磯野家の例がわかりやすいですね。
```prolog
:- op(500,xfx,<:).

サザエ <: 波平.
サザエ <: フネ.
カツオ <: 波平.
カツオ <: フネ.
わかめ <: 波平.
わかめ <: フネ.
タラオ <: マスオ.
タラオ <: サザエ.

t(A <: A).
t(A <: C) :- A <: B, t(B <: C).

:- t(タラオ <: タラオ).
:- t(タラオ <: サザエ).
:- t(タラオ <: 波平).
:- halt.
```

**Comment by hsk** [comment]
```prolog
:- op(500,xfx,<:).
t(_,A <: A).
t(Γ,A <: C):- member(A<:B,Γ), t(B<:C).
t(A <: B):- t([nat <: int,int <: rat,rat <: real],A <: B).
:- t(nat <: nat).
:- t(nat <: int).
:- t(nat <: rat).
:- t(nat <: real).
:- halt.
```
型環境に入れておくとかすればアルゴリズミックな部分型付けの規則になる。

**Comment by ryo33** [comment]
@hsk  GitHub Discussionが色々な面で不便に感じてきたのでDiscordに移行しても大丈夫ですか？

**Reply by hsk** [reply]
移動しましょうｗ
同期が取れないｗ

**Reply by ryo33** [reply]
やっぱりかなり不満に感じているのとDiscord以外を使うモチベーションがないので、移動させていただきたいです。以下が招待リンクです。
https://discord.gg/NyahPqAq9m

**Reply by hsk** [reply]
以降はdiscordを使いましょう。wikiみたいなのが欲しくなりそうな気はしますけど。
